File: app/src/main/java/com/water/broadcasesample/SmsActivity.java
package com.water.broadcasesample;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.view.View;
import android.widget.TextView;

import com.water.broadcasesample.receiver.SmsReceiver;
import com.water.utilities.PermissionHelper;

public class SmsActivity extends Activity
        implements SmsReceiver.IMessageListener {
    private TextView tvSms;
    private SmsReceiver _receiver;
    private String[] _permissions = {
            Manifest.permission.RECEIVE_SMS
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sms);
        bindViews();
    }

    // 短信权限设置
    private void bindViews() {
        PermissionHelper.checkPermission(this, _permissions);

//        Intent intent = new Intent();
//        intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
//        Uri uri = Uri.fromParts("package", getApplication().getPackageName(), null);
//        intent.setData(uri);
//        startActivity(intent);

        tvSms = (TextView) findViewById(R.id.tvSms);
        _receiver = new SmsReceiver();

//        _receiver.setOnReceivedMessageListener(
//                new SmsReceiver.IMessageListener() {
//            @Override
//            public void OnReceived(String msg) {
//
//            }
//        });
        _receiver.setOnReceivedMessageListener(this);
    }

    @Override
    public void OnReceived(String message) {
        tvSms.setVisibility(View.VISIBLE);
        tvSms.setText(message);
    }
}

File: app/src/main/kotlin/com/github/androidhappyclub/broadcasesample/SmsActivity.kt
package com.github.androidhappyclub.broadcasesample

import android.Manifest
import android.content.IntentFilter
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.ave.vastgui.tools.utils.permission.requestPermission
import com.ave.vastgui.tools.viewbinding.viewBinding
import com.github.androidhappyclub.broadcasesample.databinding.ActivitySmsBinding
import com.github.androidhappyclub.broadcasesample.log.mLogFactory
import com.github.androidhappyclub.broadcasesample.receiver.SmsReceiver
import com.github.androidhappyclub.broadcasesample.receiver.SmsReceiver.IMessageListener

class SmsActivity : AppCompatActivity(R.layout.activity_sms), IMessageListener {

    private val binding by viewBinding(ActivitySmsBinding::bind)
    private val logger = mLogFactory.getLog(SmsActivity::class.java)
    private val receiver = SmsReceiver().apply {
        setOnReceivedMessageListener(this@SmsActivity)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requestPermission(Manifest.permission.RECEIVE_SMS) {
            granted = {
                logger.d("权限 $it 已经被授予")
            }
        }

        registerReceiver(receiver, IntentFilter("android.provider.Telephony.SMS_RECEIVED"))
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterReceiver(receiver)
    }

    override fun onReceived(message: String?) {
        binding.tvSms.visibility = View.VISIBLE
        binding.tvSms.text = message
    }

}